vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package routines
+
+import (
+	"encoding/json"
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	controllerfetcher "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
+)
+
+const (
+	vpaPostProcessorReplicaRestrictionsSuffix = "replicaRestrictions"
+)
+
+// ReplicaRestrictionsPostProcessor ensures that defined ratio constraints between resources is applied.
+// The definition is done via annotation on the VPA object with format: vpa-post-processor.kubernetes.io/replicaRestrictions="[lowWatermark, highWatermark]"
+// - lowWatermark is the maximum number of replicas to allow a vertical downscale
+// - highWatermark is the minimum number of replicas to allow a vertical upscale
+// In other words:
+// - If we have less than <lowWatermark> replicas, we will allow the pods to be smaller, the intent it to keep the number
+//   of pods above <lowWatermark>. This allows you to improve resource utilisation to avoid having few very big and underutilised pods.
+// - If we have more than <highWatermark> replicas, we will allow the pods to be bigger, the intent it to keep the number
+//   of pods below <highWatermark>. This allows you to reduce the pressure on the k8s control plane.
+// If the annotation is defined and set to an empty value, the default value is used: [10, 100].
+// Any increase of resources is considered an upscale, any decrease of resources is considered a downscale.
+// This is best used with the `maintainedRatio` policy or with a single controlled resource.
+type ReplicaRestrictionsPostProcessor struct {
+	ControllerFetcher controllerfetcher.ControllerFetcher
+}
+
+// Process applies the Resource Ratio post-processing to the recommendation.
+func (r *ReplicaRestrictionsPostProcessor) Process(vpa *model.Vpa, recommendation *vpa_types.RecommendedPodResources, _ *vpa_types.PodResourcePolicy) *vpa_types.RecommendedPodResources {
+	watermarks := readWatermarksFromVPAAnnotations(vpa)
+	if watermarks == nil || len(watermarks) == 0 || recommendation == nil {
```

Comment:
probably worth checking that the array is of length 2

Reply:
done